test/e2e/storage/csi_conformance.go

```diff
@@ -0,0 +1,282 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	"context"
+	"fmt"
+	"regexp"
+
+	"k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	//	csiv1alpha1 "k8s.io/csi-api/pkg/apis/csi/v1alpha1"
+	csiclient "k8s.io/csi-api/pkg/client/clientset/versioned"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/framework/podlogs"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	//	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	"k8s.io/kubernetes/test/e2e/storage/testsuites"
+	"k8s.io/kubernetes/test/e2e/storage/utils"
+	//	imageutils "k8s.io/kubernetes/test/utils/image"
+
+	"crypto/sha256"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+// This executes testSuites for csi volumes.
+var _ = utils.SIGDescribe("CSI Conformance", func() {
+	f := framework.NewDefaultFramework("csi-conformance")
+
+	var (
+		cancel context.CancelFunc
+		cs     clientset.Interface
+		csics  csiclient.Interface
+		ns     *v1.Namespace
+		// Common configuration options for each driver.
+	)
+
+	BeforeEach(func() {
+		ctx, c := context.WithCancel(context.Background())
+		cancel = c
+		cs = f.ClientSet
+		csics = f.CSIClientSet
+		ns = f.Namespace
+		// Debugging of the following tests heavily depends on the log output
+		// of the different containers. Therefore include all of that in log
+		// files (when using --report-dir, as in the CI) or the output stream
+		// (otherwise).
+		to := podlogs.LogOutput{
+			StatusWriter: GinkgoWriter,
+		}
+		if framework.TestContext.ReportDir == "" {
+			to.LogWriter = GinkgoWriter
+		} else {
+			test := CurrentGinkgoTestDescription()
+			reg := regexp.MustCompile("[^a-zA-Z0-9_-]+")
+			// We end the prefix with a slash to ensure that all logs
+			// end up in a directory named after the current test.
+			to.LogPathPrefix = framework.TestContext.ReportDir + "/" +
+				reg.ReplaceAllString(test.FullTestText, "_") + "/"
+		}
+		podlogs.CopyAllLogs(ctx, cs, ns.Name, to)
+
+		// pod events are something that the framework already collects itself
+		// after a failed test. Logging them live is only useful for interactive
+		// debugging, not when we collect reports.
+		if framework.TestContext.ReportDir == "" {
+			podlogs.WatchPods(ctx, cs, ns.Name, GinkgoWriter)
+		}
+	})
+
+	AfterEach(func() {
+		cancel()
+	})
+
+	Context("CSI attach test using HostPath driver [Feature:CSIDriverRegistry]", func() {
```

Comment:
These test cases already exist so I don't think we need to copy them as new test suites.

Reply:
Conformance tests require to be individual tests and cannot be in table format, each with their own 'It' block without a loop that programmatically creates the tests and each of them will have their own documentation. Unfortunately the framework CSI tests have will not work here. Please advice.

Reply:
Ok, let's at least try to refactor the tests into a common function that can be shared by conformance vs non-conformance tests. But also, I don't think we should use these CSI attach tests.  They are more like integration tests and do not really test end-user behavior.

I outlined some possible test cases we can add here: https://github.com/kubernetes/kubernetes/issues/69602